File: CombatCore/Collision.cs
namespace CombatCore;

public static class Collision
{
    /// <summary>
    /// Axis-aligned box overlap test. Chosen over capsules/spheres for this
    /// slice because AABB-vs-AABB needs no sqrt and no trig - it's three
    /// independent 1D interval overlap checks - which keeps the determinism
    /// surface area as small as possible. Capsule hitboxes are a reasonable
    /// upgrade later (better fit for limbs) but would only be worth the added
    /// complexity once the frame-data authoring pipeline is mature.
    /// </summary>
    public static bool AabbOverlap(
        FixedVector3 centerA, FixedVector3 halfExtentsA,
        FixedVector3 centerB, FixedVector3 halfExtentsB)
    {
        return Fixed.Abs(centerA.X - centerB.X) <= (halfExtentsA.X + halfExtentsB.X)
            && Fixed.Abs(centerA.Y - centerB.Y) <= (halfExtentsA.Y + halfExtentsB.Y)
            && Fixed.Abs(centerA.Z - centerB.Z) <= (halfExtentsA.Z + halfExtentsB.Z);
    }

    /// <summary>
    /// Resolves a FrameBox (authored in the attacker's local, facing-relative
    /// space) into a world-space center point for this frame, given the
    /// player's current position and facing direction.
    /// </summary>
    public static FixedVector3 ResolveWorldCenter(FixedVector3 playerPosition, int facingSign, FrameBox box)
    {
        return new FixedVector3(
            playerPosition.X + box.LocalOffset.X * facingSign,
            playerPosition.Y + box.LocalOffset.Y,
            playerPosition.Z + box.LocalOffset.Z);
    }
}

File: Client/src/sim/Collision.ts
import { Fixed } from "./Fixed.ts";
import { FixedVector3 } from "./FixedVector3.ts";
import type { FrameBox } from "./FrameData.ts";

export function aabbOverlap(
  centerA: FixedVector3, halfExtentsA: FixedVector3,
  centerB: FixedVector3, halfExtentsB: FixedVector3,
): boolean {
  return (
    Fixed.abs(Fixed.sub(centerA.x, centerB.x)).raw <= Fixed.add(halfExtentsA.x, halfExtentsB.x).raw &&
    Fixed.abs(Fixed.sub(centerA.y, centerB.y)).raw <= Fixed.add(halfExtentsA.y, halfExtentsB.y).raw &&
    Fixed.abs(Fixed.sub(centerA.z, centerB.z)).raw <= Fixed.add(halfExtentsA.z, halfExtentsB.z).raw
  );
}

export function resolveWorldCenter(playerPosition: FixedVector3, facingSign: number, box: FrameBox): FixedVector3 {
  return new FixedVector3(
    Fixed.add(playerPosition.x, Fixed.mulInt(box.localOffset.x, facingSign)),
    Fixed.add(playerPosition.y, box.localOffset.y),
    Fixed.add(playerPosition.z, box.localOffset.z),
  );
}
